head/core/can_task.cpp

```diff
@@ -0,0 +1,199 @@
+
+/** Dispatcher to the various right motor handlers */
+static auto dispatcher_right_motor = can_dispatch::Dispatcher(
+    check_for_node_id_right, motor_dispatch_target_right,
+    motion_dispatch_target_right, move_group_dispatch_target_right,
+    device_info_dispatch_target_right);
+
+/** Dispatcher to the various left motor handlers */
+static auto dispatcher_left_motor = can_dispatch::Dispatcher(
+    check_for_node_id_left, motor_dispatch_target_left,
+    motion_dispatch_target_left, move_group_dispatch_target_left,
+    device_info_dispatch_target_left);
+
+static auto main_dispatcher =
+    can_dispatch::Dispatcher([](auto _) -> bool { return true; },
+                             dispatcher_right_motor, dispatcher_left_motor);
+
+/**
+ * The type of the message buffer populated by HAL ISR.
+ */
+static auto read_can_message_buffer =
+    freertos_message_buffer::FreeRTOSMessageBuffer<1024>{};
+static auto read_can_message_buffer_writer =
+    can_message_buffer::CanMessageBufferWriter(read_can_message_buffer);
+
+/**
+ * New CAN message callback.
+ *
+ * @param identifier Arbitration id
+ * @param data Message data
+ * @param length Message data length
+ */
+void callback(void* cb_data, uint32_t identifier, uint8_t* data,
+              uint8_t length) {
+    read_can_message_buffer_writer.send_from_isr(identifier, data,
+                                                 data + length);  // NOLINT
+}
+
+/**
+ * Entry point for the reader task.
+ * TODO (2021-12-15, AL): Most of what happens in this task should be moved out
+ *  when we move to separate motor tasks.
+ */
+[[noreturn]] void can_task::CanMessageReaderTask::operator()(
+    can_bus::CanBus* can_bus) {
+    can_bus->set_incoming_message_callback(nullptr, callback);
+
+    auto filter = can_arbitration_id::ArbitrationId();
+
+    // Accept broadcast
+    filter.node_id(can_ids::NodeId::broadcast);
+    can_bus->add_filter(CanFilterType::mask, CanFilterConfig::to_fifo0, filter,
+                        can_arbitration_id::ArbitrationId::node_id_bit_mask);
+
+    // Accept any head
+    filter.node_id(can_ids::NodeId::head);
+    can_bus->add_filter(CanFilterType::mask, CanFilterConfig::to_fifo1, filter,
+                        can_arbitration_id::ArbitrationId::node_id_bit_mask);
+
+    // Accept head right
+    filter.node_id(can_ids::NodeId::head_r);
+    can_bus->add_filter(CanFilterType::mask, CanFilterConfig::to_fifo1, filter,
+                        can_arbitration_id::ArbitrationId::node_id_bit_mask);
+
+    // Accept head left
+    filter.node_id(can_ids::NodeId::head_l);
+    can_bus->add_filter(CanFilterType::mask, CanFilterConfig::to_fifo1, filter,
+                        can_arbitration_id::ArbitrationId::node_id_bit_mask);
+
+    // Reject everything else.
+    can_bus->add_filter(CanFilterType::mask, CanFilterConfig::reject, 0, 0);
+
+    // TODO (al, 2021-12-21): Must account for this!!
+    //    motor_left.driver.setup();
```

Comment:
noting this quick - is this a leftover comment or are these in fact never called?

Reply:
It should be deleted.